### PR TITLE
Add privileged-context distillation to GOLD

### LINKED ISSUE #5741: Can gold_trainer support privileged context?
Current gold_trainer(trl/experimental/gold/gold_trainer.py) seems to only support distillation between students and teachers that have the same input. I wonder whether it can support teacher model with privileged context as in sdft_trainer.

### PR DESCRIPTION
**Pull request by eshwanthkartitr**
# Introduce privileged-context distillation for GOLDTrainer

This PR introduces an opt-in privileged-context distillation path for `GOLDTrainer`, allowing the teacher model to see per-example privileged context (like reasoning traces or extra instructions) while the student's prompt remains unchanged.

### Changes made:
* **Configuration**: Added `use_privileged_context`, `teacher_prompt_template`, and `privileged_context_column` to `GOLDConfig` with loud validation to ensure incompatible arguments (e.g., `seq_kd`, `use_uld_loss`, `lmbda > 0`, Liger/VLM combos) are rejected upfront.
* **Dataset Preparation**: Modified `_prepare_dataset_with_original_text` and `DataCollatorForChatML` to build and preserve a `teacher_prompt_text` field. 
* **Forward Pass Interception**: Implemented `build_teacher_inputs_from_completion_ids` to concatenate the generated `teacher_prompt_text` with the student's byte-identical completion token IDs, allowing a split forward pass for the JS divergence loss calculation.
* **Documentation & Tests**: Updated `docs/source/gold_trainer.md` with usage examples and added 6 new test cases to rigorously validate context injection and mathematical token equivalence.
* **Fix**: Added a minor MPS dtype fallback (`torch.float32`) for a tensor creation edge case during logging on Apple Silicon.

Fixes #5741 

## Before submitting

- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@kashif @lewtun @qgallouedec

---

> [!NOTE]
> **Medium Risk**
> Changes GOLD training loss and dataset preparation on an experimental trainer path, but the feature is gated, validated against incompatible options, and covered by new tests; teacher prompts are uncapped so context-length mistakes are an operational footgun.
> 
> **Overview**
> Adds **opt-in privileged-context distillation** to `GOLDTrainer`: the student still trains on the normal prompt and completion, while the teacher is scored on a separate prompt that injects per-example extra context (default `"{prompt}\n\n{privileged_context}"`).
> 
> **Configuration** on `GOLDConfig`: `use_privileged_context`, `teacher_prompt_template`, and `privileged_context_column`, with upfront validation rejecting incompatible modes (on-policy `lmbda > 0`, `seq_kd`, ULD, Liger, packing, VLMs).
> 
> **Training path**: dataset prep builds `teacher_prompt_text` (rewriting the last user turn for chat data); the ChatML collator carries it through; `compute_loss` tokenizes the enriched teacher prompt and appends the **same completion token IDs** as the student via `build_teacher_inputs_from_completion_ids`, then runs standard JSD. Teacher prompts are not truncated to `max_length`.
> 
> Docs cover usage and limits. Tests cover config validation, prompt building, token alignment, and loss wiring. Also fixes distributed logging tensor dtype on MPS (`float32` instead of `float64`).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4c28b0c9e00bf5e952cb3f3edc95b4d1405f5178. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/gold/gold_trainer.py [hunk 1]
```diff
@@ -1993,14 +2031,64 @@ def _prepare_dataset(
...
+        """Build the teacher prompt by enriching the final user turn with privileged context."""
+        if privileged_context is None:
+            raise ValueError(
+                f"`{privileged_context_column}` must not be None when `use_privileged_context=True`."
+            )
+
+        privileged_context = str(privileged_context)
+        if isinstance(prompt, list):
+            messages = [dict(message) for message in prompt]
+            last_message = messages[-1]
```
[the remark this answers is not in the document]
Comment (eshwanthkartitr) [reply trl/experimental/gold/gold_trainer.py:2089]: addressed in the latest pr with a check

**Comment by eshwanthkartitr** [comment]
@cmpatino PTAL

**Comment by qgallouedec** [comment]
Hey @eshwanthkartitr, thanks for turning this around so quickly, the scope matches what we agreed on in #5741.

Before I do a full review, CI needs to go green: `Tests (experimental)` has 12 failures, and 11 of them share one root cause: `GOLDTrainer.__init__` now reads `args.use_privileged_context` / `args.teacher_prompt_template` / `args.privileged_context_column` unguarded, and `_prepare_dataset_with_original_text` reads `self.use_privileged_context`, but many existing tests hand-build `args` as a `SimpleNamespace` or instantiate via `GOLDTrainer.__new__(GOLDTrainer)`

Ping one of usonce it's green and we'll review properly :)

**Comment by eshwanthkartitr** [comment]
Hi @qgallouedec I've pushed a commit that addresses the reported issue, and the changes are passing locally. Could you please approve/trigger the workflow so the CI can run the code quality checks and test suite? Once it completes, I'll verify that everything is green. Thanks!

File: trl/experimental/gold/gold_config.py [hunk 2]
```diff
@@ -64,6 +64,14 @@ class GOLDConfig(SFTConfig):
         seq_kd (`bool`, *optional*, defaults to `False`):
             Seq_kd parameter that controls whether to perform Sequence-Level KD (can be viewed as supervised FT on
             teacher-generated output).
+        use_privileged_context (`bool`, *optional*, defaults to `False`):
+            Whether to condition the teacher on per-example privileged context while leaving the student prompt
+            unchanged. This is supported only for text, off-policy standard JSD distillation.
+        teacher_prompt_template (`str` or `None`, *optional*, defaults to `None`):
+            Template used to construct the teacher prompt. It must contain `{prompt}` and may contain
+            `{privileged_context}`. When `None`, defaults to `"{prompt}\n\n{privileged_context}"`.
```
[the remark this answers is not in the document]
Comment (cmpatino) [reply trl/experimental/gold/gold_config.py:72]: Please address this comment

File: trl/experimental/gold/gold_trainer.py [hunk 3]
```diff
@@ -2825,7 +3011,7 @@ def log(self, logs: dict[str, float], start_time: float | None = None) -> None:
                     self._matched_step_eq,
                     self._unmatched_step_eq,
                 ],
-                dtype=torch.float64,
+                dtype=torch.float64 if device.type != "mps" else torch.float32,
```
Comment (cmpatino) [inline trl/experimental/gold/gold_trainer.py:3014]: That's a good catch! 

I agree with this fix, but I think it's out of scope for this PR. Also, the issue is present in other trainers related to distillation. Would you be open to opening a PR with the same fix for the GOLD and IW-OPD trainers?

File: trl/experimental/gold/gold_trainer.py [hunk 4]
```diff
@@ -2187,14 +2321,19 @@ def tokenize_with_original_text(example, processing_class, dataset_text_field, m
...
+                    result["teacher_prompt_text"] = teacher_prompt_text
                 return result
 
             dataset = dataset.map(
                 tokenize_with_original_text,
                 fn_kwargs={
                     "processing_class": processing_class,
                     "dataset_text_field": args.dataset_text_field,
-                    "max_length": args.max_length,
+                    "max_length": getattr(args, "max_length", None),
```
[the remark this answers is not in the document]
Comment (cmpatino) [reply trl/experimental/gold/gold_trainer.py:2345]: I agree with this comment, but I'm curious why you marked it as resolved.

**Review by cmpatino — COMMENTED** [review]
The PR is in very good shape. Thank you for addressing the comments about the CI.

I left three minor comments that should hopefully be easy to fix. After those, I'll approve the PR.
